### PR TITLE
feat: custom properties for deployer

### PR DESCRIPTION
**Pull request by kakhaUrigashvili**
Addressing the feature request: https://github.com/alexa/ask-cli/issues/211
Example:
```
...
 "userConfig": {
          "runtime": "nodejs10.x",
          "handler": "index.handler",
          "templatePath": "./infrastructure/cfn-deployer/skill-stack.yaml",
          "awsRegion": "us-east-1",
          "templateParameters": {
            "ParamOne": "foo",
            "ParamTwo": "bar"
          }
        }
...
```

Custom params can overwrite our reserved params.

### CODE REVIEW

File: lib/builtins/deploy-delegates/cfn-deployer/helper.js [hunk 1]
```diff
@@ -99,6 +100,11 @@ function deployStack(reporter, options, callback) {
             }
         ]
     };
+
+    // merge custom deployer parameters with standard parameters
+    // custom deployer parameters can overwrite standard parameters
+    stackOptions.stackParameters = R.unionWith(R.eqBy(R.prop('ParameterKey')), deployerParameters, stackOptions.stackParameters);
```
Comment (kakhaUrigashvili) [inline lib/builtins/deploy-delegates/cfn-deployer/helper.js:106]: if we dont want standard params to be overwritten, we should switch the order of union.

**Comment by conman124** [comment]
Would it be possible to also allow passing parameters in the skillInfrastructure section of ask-resources.json?

**Comment by kakhaUrigashvili** [comment]
> Would it be possible to also allow passing parameters in the skillInfrastructure section of ask-resources.json?

agree. I will update it.

File: docs/concepts/Alexa-Skill-Project-Definition.md [hunk 2]
```diff
@@ -60,7 +60,11 @@ Below shows an example config JSON file, using `@ask-cli/cfn-deployer` deployer
         "userConfig": {                     // tracks settings for the deployer, deployer-specific
           "runtime": "{lambdaRuntime}",
           "handler": "{lambdaHandler}",
-          "templatePath": "stack.yaml"
+          "templatePath": "stack.yaml",
+          "templateParameters": { // custom parameters to pass to the deployer. Supported for CloudFormation deployer only
```
Comment (RonWang) [inline docs/concepts/Alexa-Skill-Project-Definition.md:64]: More correct wording: `Supported for @ask-cli/cfn-deployer only`

File: lib/builtins/deploy-delegates/cfn-deployer/helper.js [hunk 3]
```diff
@@ -99,6 +100,9 @@ function deployStack(reporter, options, callback) {
             }
         ]
     };
+
+    _mergeTemplateParameters(stackOptions.stackParameters, templateParameters);
```
Comment (RonWang) [inline lib/builtins/deploy-delegates/cfn-deployer/helper.js:104]: I'm always thinking that current flow is in need of more transparency regarding the input parameters to CFN stack; skillId and codeBucket/Key/Version are values passed from previous skillMetadata and S3 deployment, only lambdaRuntime and lambdaHandler are the two required values that we read from userConfig, with a special logic to map `userConfig.runtime` and `userConfig.handler` to the two stack parameters. What if we put them as `userConfig.templateParameter.lambdaRuntime` and `userConfig.templateParameter.lambdaHandler`, and use a **$ref** syntax to have a pass-through experiences for stack parameters? Please check this example:

```
 "userConfig": {
    "awsRegion": "us-east-1",
    "templatePath": "./template/stack.yaml"
    "templateParameters": { 
        "LambdaHandler": "nodejs10.x",
        "LambdaRuntime": "index.handler",
        "SkillId": "$state.skillId",
        "CodeBucket": "$state.@ask-cli/cfn-deployer.deployState.s3.bucket",
        "CodeKey": "$state.@ask-cli/cfn-deployer.deployState.s3.key",
        "CodeVersion": "$state.@ask-cli/cfn-deployer.deployState.s3.objectVersion",
        "OtherParameterIfExistInStack": "..."
    }
},
```

In this way, CLI can be a totally pass-through bridge for all the parameters of cfn-stack.

File: lib/builtins/deploy-delegates/cfn-deployer/helper.js [hunk 4]
```diff
@@ -133,6 +137,18 @@ function deployStack(reporter, options, callback) {
...
+    Object.keys(templateParameters).forEach(ParameterKey => {
+        const ParameterValue = templateParameters[ParameterKey];
+        const index = stackParameters.findIndex(p => p.ParameterKey === ParameterKey);
+        if (index > -1) {
+            stackParameters[index].ParameterValue = ParameterValue;
+        } else {
+            stackParameters.push({ ParameterKey, ParameterValue });
+        }
+    });
+}
```
Comment (RonWang) [inline lib/builtins/deploy-delegates/cfn-deployer/helper.js:150]: Can we use https://ramdajs.com/docs/#mergeDeepLeft to do the (deep) merge?

Comment (kakhaUrigashvili) [reply lib/builtins/deploy-delegates/cfn-deployer/helper.js:150]: merge deep left merges on property keys. we are merging arrays. Tried and it did not work as expected:
```
R.mergeDeepLeft([{ParameterKey: 'z', ParameterValue: 'f'}, {ParameterKey: 'x', ParameterValue: 'y'}],
[{ParameterKey: 'z', ParameterValue: 'r'}, {ParameterKey: 'q', ParameterValue: 'y'}]);
```

File: lib/builtins/deploy-delegates/cfn-deployer/helper.js (hunk 3 above)
Comment (kakhaUrigashvili) [reply lib/builtins/deploy-delegates/cfn-deployer/helper.js:104]: @RonWang would this be a breaking change?
When new project is generated, userConfig section looks like
```
"userConfig": {
          "runtime": "nodejs10.x",
          "handler": "index.handler",
          "templatePath": "./infrastructure/cfn-deployer/skill-stack.yaml",
          "awsRegion": "us-east-1"
        }
```
runtime and handler will be moved under templateParameters and renamed. I think this is a breaking change?

**Comment by crackmac** [comment]
Many folks would love to see this merged in or another solution. This would close the gap between skill and dependent AWS resources.

**Comment by kakhaUrigashvili** [comment]
> Many folks would love to see this merged in or another solution. This would close the gap between skill and dependent AWS resources.

@RonWang any thoughts? I am fine with the PR as it is.

**Comment by crackmac** [comment]
We also need a way to pass the cloudformation `--capabilities` flag in order to enable useful Transform macros (ie. String). 

Perhaps this should be another Feature request.

ref:
https://awscli.amazonaws.com/v2/documentation/api/latest/reference/cloudformation/create-stack.html
https://github.com/awslabs/aws-cloudformation-templates/tree/master/aws/services/CloudFormation/MacrosExamples/StringFunctions

**Comment by kakhaUrigashvili** [comment]
Closing in favor of https://github.com/alexa/ask-cli/pull/317. Please provide feedback on https://github.com/alexa/ask-cli/pull/317

**Pull request closed without merging**